src/components/legend/handle_click.js

```diff
@@ -245,17 +278,21 @@ module.exports = function handleClick(g, gd, numClicks) {
         // values should be explicitly undefined for them to get properly culled
         // as updates and not accidentally reset to the default value. This fills
         // out sparse arrays with the required number of undefined values:
-        keys = Object.keys(attrUpdate);
+        keys = Object.keys(dataUpdate);
         for(i = 0; i < keys.length; i++) {
             key = keys[i];
-            for(j = 0; j < attrIndices.length; j++) {
+            for(j = 0; j < dataIndices.length; j++) {
                 // Use hasOwnProperty to protect against falsy values:
-                if(!attrUpdate[key].hasOwnProperty(j)) {
-                    attrUpdate[key][j] = undefined;
+                if(!dataUpdate[key].hasOwnProperty(j)) {
+                    dataUpdate[key][j] = undefined;
                 }
             }
         }
 
-        Registry.call('_guiRestyle', gd, attrUpdate, attrIndices);
+        Registry.call('_guiRestyle', gd, dataUpdate, dataIndices).then(function() {
+            if(shapesUpdated) {
+                Registry.call('_guiRelayout', gd, {shapes: updatedShapes});
+            }
+        });
```

Comment:
Can we turn this into a single `_guiUpdate` call?

Reply:
Good call. Addressed in [759dbb4](https://github.com/plotly/plotly.js/pull/6653/commits/759dbb4c8d9b83834ef48d805deeee43b4f88c49).